firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteDocumentOverlayCache.java

```diff
@@ -47,7 +56,54 @@ public Overlay getOverlay(DocumentKey key) {
             "SELECT overlay_mutation, largest_batch_id FROM document_overlays "
                 + "WHERE uid = ? AND collection_path = ? AND document_id = ?")
         .binding(uid, collectionPath, documentId)
-        .firstValue(this::decodeOverlay);
+        .firstValue(row -> this.decodeOverlay(row.getBlob(0), row.getInt(1)));
+  }
+
+  @Override
+  public Map<DocumentKey, Overlay> getOverlays(SortedSet<DocumentKey> keys) {
+    hardAssert(keys.comparator() == null, "getOverlays() requires natural order");
+    Map<DocumentKey, Overlay> result = new HashMap<>();
+
+    BackgroundQueue backgroundQueue = new BackgroundQueue();
+    ResourcePath currentCollection = ResourcePath.EMPTY;
+    List<Object> accumulatedDocumentIds = new ArrayList<>();
+    for (DocumentKey key : keys) {
+      if (!currentCollection.equals(key.getCollectionPath())) {
+        processSingleCollection(result, backgroundQueue, currentCollection, accumulatedDocumentIds);
+        currentCollection = key.getCollectionPath();
+        accumulatedDocumentIds.clear();
```

Comment:
I think you need to assign `accumulatedDocumentIds` to a new object rather than clearing it. This list is specified to `SQLitePersistence.LongQuery` which holds an iterator to its contents. And modifying the list underneath a live iterator is undefined.

Reply:
That is a quality code review right here. Thanks for catching this.

Reply:
Actually - I think it might have been fine since all processing happens synchronously. But it is more clear with the suggested fix.